README.md
# IDAVSCode

在 VSCode 中调试 IDAPython!

参考了[idacode](https://github.com/ioncodes/idacode)

<p align="center">
  [中文]
  [<a href="doc/README-en.md">English</a>]
</p>

## Features

![](idavscode/image/demo.webp)

- 在 VSCode 中按下 <kbd>F5</kbd> 即可开始调试！
- 调试时，可以直接通过 VSCode 中的 `args` 设置 `idc.ARGV`
- Hook 了调试环境中的 `ida_pro.qexit, idc.qexit`，调试环境、生产环境无需额外判断


## Requirement


1. `python >= 3.10` # 理论上替换掉 `match case` 语句应该就可以3.x全兼容
3. `python -m pip install debugpy tornado`
2. [Python extension for VSCode](https://marketplace.visualstudio.com/items?itemName=ms-python.python)


## Install

1. 下载 IDA 插件，并解压到 Plugin 文件夹；
2. 安装 [Vscode 拓展](https://marketplace.visualstudio.com/items?itemName=Cirn09.idavscode)

## launch.json

```json
{
    "name": "IDAPython debug",
    "type": "idapython",
    "request": "launch",
    "program": "${file}",
    "host": "localhost",                                // control hostname
                                                        // 和 IDA 端一致
    "port": 5677,                                       // control port
                                                        // 和 IDA 端一致
    "timeout": 3000,                                    // 连接失败超时
    "pythonPath": "${command:python.interpreterPath}",  // python path (IDA used)
                                                        // VSCode 和 IDA 使用同版本的话就不用管这字段了
    "logFile": "",                                      // debug log file
    "debugConfig": {                                    // Python extension debug config
                                                        // 传递给 Vscode Python 拓展的调试配置
        "type": "python",
        "request": "attach",
        "justMyCode": true,
        "connect": {
            // "host": "localhost",                     // optional, default as seam as control host
                                                        // 可选，不填的话默认用上层的"host"
            "port": 5678                                // debug port
        },
        "cwd": "${workspaceFolder}"
    }
}
```

## 原理

IDA 端插件启动两个服务，一个控制服务、一个调试服务([debugpy](https://github.com/microsoft/debugpy))。

VSCode 端拓展拦截调试任务，将调试上下文信息发送到 IDA 控制服务端，IDA 控制服务负责启动调试服务，并按照上下文运行目标脚本。准备完成后，VSCode 拓展将调试任务替换为 Python Remote Attach 任务。脚本运行结束后，VSCode 拓展主动停止调试。

## 已知缺陷和待办

该插件诸多的缺陷、不如意大都因为这是一个 IDAPython 插件，它和被调试脚本运行在同一个 Python 实例下，某种意义上说它是自己调试自己。

所以解决以下以及其他潜在缺陷最好的方案是使用 C++ 重构。

- [x] `idc.ARGV`

- [x] 取消调试/连接超时

- [x] 被调试脚本触发异常后，无法正常再次调试。

    应该和IDAPython线程有关，控制和调试服务都是运行在子线程上，被调试脚本通过 `ida_kernwin.execute_sync` 运行在主线程。添加 `ida_kernwin.refresh_idaview_anyway()` 似乎就正常了。

- [ ] 对由被调试脚本 import 的模块的修改不会生效。

    https://github.com/Cirn09/idavscode/issues/1

    现在是一个不完美的解决方案。

- [ ] 调试服务无法终止：调试服务使用了 [debugpy](https://github.com/microsoft/debugpy)，debugpy 目前只提供了启动服务的接口，没有停止服务的接口 ([相关issue](https://github.com/microsoft/debugpy/issues/870))。所以调试服务启动后，debugpy会一直占用调试端口。

    暂时的解决方案：可以手动重启 IDA，或者结束 IDA 的 Python 子进程（这个子进程是debugpy启动的，IDAPython 使用 Python.dll，不会启动单独的 Python 进程）

- [x] 进行过一次调试、关闭 idb/i64、重新打开新 idb/i64 后无法正常调试（无法设置断点，stdout没有正确转发）。

    还是 debugpy 没有提供关闭接口问题的延续，关闭 idb 时 ida 会卸载插件，重开又会重新加载。

    添加 `PLUGIN_FIX` 属性，每次调试前都调用 `debugpy.debug_this_thread()` 即可解决。

- [ ] （可能存在）Python调试设置不生效：VSCode 插件拦截替换调试任务使用的是 `vscode.DebugConfigurationProvider { resolveDebugConfiguration, resolveDebugConfigurationWithSubstitutedVariables }` 两个接口，其中使用 `resolveDebugConfiguration` 补充调试配置、使用 `resolveDebugConfigurationWithSubstitutedVariables` 替换调试任务。

    实际观察，通过 `resolveDebugConfigurationWithSubstitutedVariables` 替换任务的效果和调用 `vscode.debug.startDebugging` 的效果不完全一致。抓包发现，前者创建的调试 session 少了一些选项，推测是这种方法绕过了一些 `launch.json -> session config` 流程。

- [ ] `launch.json` 中的 `pythonPath` 会被自动修正为 `python`，不确定是我这边才有的问题还是普遍存在。
- [ ] 图标
- [ ] 测试
- [ ] README-en.md
- [ ] 切分流程，提供更细粒度的命令
- [ ] 改一改是不是能做个 GDBPython 调试助手，或者是嵌入 Python 调试助手


idavscode/CHANGELOG.md
# Change Log

## 0.0.5 - 2023.03.20

### Bug Fixes

- Sometimes debugging doesn't auto stop after script stop
- "timeout" in launch.json

## 0.0.4 - 2023.03.15

### Added

- Support `idc.ARGV`
- Adding connect timeout

## 0.0.3 - 2022.07.26

### Changed

- Better Python file encoding handling

### Fixed

- Python module cache cause modify of module not effect
- Can't debug after reopen idb/i64
- Python syntax error handling

## 0.0.2
### Fixed

- Can't debug again after python exception

## 0.0.1 - 2022.07.13

- Initial release

ida/idavsc.py
import os
import ida_kernwin, idaapi

from dbg_server import Config, Server, DebugServerCannotStopError

CONFIG_FILE = os.path.join(idaapi.get_user_idadir(), 'config.cfg')


# =====================================================================================
# Core
# =====================================================================================
class Plugin(idaapi.plugin_t):

    def __init__(self) -> None:
        self.comment = "IDAPython VSCode Debugger"
        # PLUGIN_HIDE: do not show this plugin in the Edit->Plugins menu
        # PLUGIN_FIX: keep plugin loaded until IDA stops. because have no way to stop debugpy
        self.flags = idaapi.PLUGIN_HIDE | idaapi.PLUGIN_FIX
        self.wanted_name = "IDAVSC"
        if not os.path.exists(CONFIG_FILE):
            self.config = Config()
        else:
            self.config = Config.from_file(CONFIG_FILE)
        self.server = Server(self.config)

    @property
    def running(self):
        return self.server.running

    def init(self):
        """
        This is called by IDA when it is loading the plugin.
        """
        StartMenuHandle.register(self)
        StopMenuHandle.register(self)
        OptionMenuHandle.register(self)

        return idaapi.PLUGIN_KEEP

    def run(self, arg):
        """
        This is called by IDA when this file is loaded as a script.
        """
        pass

    def term(self):
        """
        This is called by IDA when it is unloading the plugin.
        """
        self.config.save(CONFIG_FILE)
        self.stop(True)


    def start(self):
        self.server.start()
        print('[IDAVSC] Server started')

    def stop(self, slient=False):
        try:
            self.server.stop()
            print('[IDAVSC] Server stopped')
        except DebugServerCannotStopError:
            if not slient:
                dialog = ErrorDialog(
                    'Control server stoped, but debug server is still running and cannot be stopped currently.\nCheck here for more: https://github.com/microsoft/debugpy/issues/870'
                )
                dialog.Execute()
                dialog.Free()
            

    def option(self):
        dialog = OptionDialog(self.config.host, self.config.port)
        if dialog.Execute() == 1:
            self.config.host = dialog.host
            self.config.port = dialog.port
            self.config.save(CONFIG_FILE)
        dialog.Free()


# =====================================================================================
# UI
# =====================================================================================
class MenuHandle(ida_kernwin.action_handler_t):
    NAME = ''
    TEXT = ''
    TOOLTIP = ''
    HOTKEY = None
    PATH = ''

    def __init__(self, plugin: Plugin) -> None:
        super(MenuHandle, self).__init__()
        self.plugin = plugin

    @classmethod
    def register(self, plugin: Plugin) -> None:
        desc = ida_kernwin.action_desc_t(self.NAME, self.TEXT, self(plugin),
                                         self.HOTKEY, self.TOOLTIP)
        if not ida_kernwin.register_action(desc):
            print(f'Failed to register action: {self.NAME}')
        ida_kernwin.attach_action_to_menu(self.PATH, self.NAME,
                                          ida_kernwin.SETMENU_APP)


class StartMenuHandle(MenuHandle):

    NAME = 'idavsc:start'
    TEXT = "Start"
    TOOLTIP = "Start debug server"
    PATH = 'Edit/IDAVSC/Start'

    def activate(self, ctx):
        self.plugin.start()

    def update(self, ctx):
        if self.plugin.running:
            return ida_kernwin.AST_DISABLE
        else:
            return ida_kernwin.AST_ENABLE


class StopMenuHandle(MenuHandle):
    NAME = 'idavsc:stop'
    TEXT = "Stop"
    TOOLTIP = "Stop debug server"
    PATH = 'Edit/IDAVSC/Stop'

    def activate(self, ctx):
        self.plugin.stop()

    def update(self, ctx):
        if self.plugin.running:
            return ida_kernwin.AST_ENABLE
        else:
            return ida_kernwin.AST_DISABLE


class OptionMenuHandle(MenuHandle):
    NAME = 'patching:option'
    TEXT = "Option..."
    TOOLTIP = "Option"
    PATH = 'Edit/IDAVSC/Option...'

    def activate(self, ctx):
        self.plugin.option()

    def update(self, ctx):
        return ida_kernwin.AST_ENABLE_ALWAYS


class OptionDialog(ida_kernwin.Form):

    def __init__(self, host, control_port):
        super(OptionDialog, self).__init__(
            r"""STARTITEM 0
BUTTON YES* OK
IDACODE :: OPTION

<Hostname       :{c_host}>
<Port           :{c_port}>
            """, {
                'c_host': self.StringInput(value=host),
                'c_port': self.NumericInput(value=control_port, tp=self.FT_DEC),
            })

        self.Compile()

    @property
    def host(self):
        return self.c_host.value

    @property
    def port(self):
        return self.c_port.value


class ErrorDialog(ida_kernwin.Form):

    def __init__(self, text):
        super(ErrorDialog, self).__init__(
            r"""STARTITEM 1
BUTTON YES* OK
IDACODE :: Error

{c_text}
            """, {
                'c_text': self.StringLabel(value=text),
            })

        self.Compile()


def PLUGIN_ENTRY():
    return Plugin()


ida/dbg_server/utils.py
import re, os, sys

import idc
import ida_pro
import ida_kernwin

PYTHON_DEFAULT_ENCODING = os.getenv("PYTHONIOENCODING", "utf-8")


class UnknowEncodingError(Exception):
    pass


class PythonFile(object):
    def __init__(self, path: str, cwd=None, argv=[], env={}, encoding=None):
        self.path = path
        self.cwd = cwd
        self.argv = [self.path]
        self.argv.extend(argv)
        self.env = env
        self.encoding = encoding

        if os.path.isabs(self.path):
            self.abspath = self.path
        else:
            self.abspath = os.path.join(self.cwd, self.path)

        self.globals = {"__name__": "__main__", "__file__": self.abspath}

        self.compile()

    def compile(self):
        if self.encoding is None:
            with open(self.abspath, "rb") as f:
                raw = f.read()

            try:
                code_text = raw.decode(PYTHON_DEFAULT_ENCODING)
                self.encoding = PYTHON_DEFAULT_ENCODING
            except UnicodeDecodeError:
                encoding_pat = re.compile(r"\s*#.*coding[:=]\s*([-\w.]+).*")
                for line in raw.decode(self.encoding, errors="replace").split("\n"):
                    match = encoding_pat.match(line)
                    if match:
                        self.encoding = match.group(1)
                        break

                if self.encoding is None:
                    raise UnknowEncodingError("unknow file encoding")
                code_text = raw.decode(self.encoding)
        else:
            with open(self.abspath, "r", encoding=self.encoding) as f:
                code_text = f.read()

        self.code = compile(code_text, self.abspath, "exec", optimize=0)

    def _before_exec(self):
        # patch
        self._orig_argv = sys.argv
        sys.argv = [self.path]
        self._orig_idc_argv = idc.ARGV
        idc.ARGV = self.argv
        self._orig_env = os.environ.copy()
        os.environ.update(self.env)
        self._orig_cwd = os.getcwd()
        os.chdir(self.cwd)
        self._idp_qexit = ida_pro.qexit
        ida_pro.qexit = lambda *_: None
        self._idc_qexit = idc.qexit
        idc.qexit = lambda *_: None
        self._orig_modules = sys.modules.copy()

        prapare_code = """
import debugpy
debugpy.debug_this_thread()
debugpy.wait_for_client()

import ida_kernwin
ida_kernwin.refresh_idaview_anyway()
        """
        prapare = compile(prapare_code, "", "exec")

        ida_kernwin.execute_sync(lambda: exec(prapare), ida_kernwin.MFF_WRITE)

    def exec(self):
        self._before_exec()
        ida_kernwin.execute_sync(
            lambda: exec(self.code, self.globals), ida_kernwin.MFF_WRITE
        )
        self._after_exec()

    def _after_exec(self):
        # restore
        sys.argv = self._orig_argv
        idc.ARGV = self._orig_idc_argv
        os.environ = self._orig_env
        os.chdir(self._orig_cwd)
        ida_pro.qexit = self._idp_qexit
        idc.qexit = self._idc_qexit
        for k in list(sys.modules.keys()):
            if k not in self._orig_modules:
                # https://docs.python.org/3/reference/import.html#the-module-cache
                del sys.modules[k]


# https://docs.python.org/3.10/library/sys.html?highlight=sys%20path#sys.path
# > As initialized upon program startup, the first item of this list,
# > path[0], is the directory containing the script that was used to
# > invoke the Python interpreter.
# > If path[0] is the **empty string**, which directs Python to search modules 
# > in the current directory first.
if "" not in sys.path:
    sys.path.insert(0, "")


ida/dbg_server/server.py
import sys
import asyncio
import json
import threading
import debugpy
import tornado.httpserver, tornado.web, tornado.websocket

from .config import Config
from .utils import PythonFile

class MessageType(object):
    StartDebugServer = 'startDebugServer'
    StopDebugServer = 'stopDebugServer'
    StopServer = 'stopServer'
    ExecuteScript = 'executeScript'

    ServerReady = 'serverReady'
    DebugServerReady = 'debugServerReady'
    DebugFinished = 'debugFinished'
    Error = 'error'

dbgsrv_running = False

class DebugServerCannotStopError(Exception):
    pass


class _Server(tornado.websocket.WebSocketHandler):

    def __init__(self, *args, **kwargs):
        super(_Server, self).__init__(*args, **kwargs)

    def on_message(self, message: str):
        global dbgsrv_running
        msg: dict = json.loads(message)
        try:
            
            match msg['type']:
                case MessageType.StartDebugServer:
                    if not dbgsrv_running:
                        host = msg['host']
                        port = msg['port']
                        logfile = msg['logfile']
                        python_path = msg['pythonPath']

                        if logfile:
                            debugpy.log_to(logfile)
                            print(f'[VSC] Debug log will be saved to {logfile}')
                        if python_path:
                            debugpy.configure({ "python": python_path })
                            print(f'[VSC] Python path set to {python_path}')

                        debugpy.listen((host, port))
                        print(f'[VSC] Debug server started on {host}:{port}')
                        dbgsrv_running = True
                    self.write_message({'type': MessageType.DebugServerReady})
                    
                case MessageType.StopDebugServer:
                    # https://github.com/microsoft/debugpy/issues/870
                    self.write_message({'type': MessageType.Error, 'info': 'Not implemented'})
                    raise NotImplementedError
                case MessageType.StopServer:
                    self.write_message({'type': MessageType.ServerReady})
                    print('[VSC] Server stopped')
                    self.close()
                case MessageType.ExecuteScript:
                    if not dbgsrv_running:
                        self.write_message({'type': MessageType.Error, 'info': 'debug server is not running'})
                        return

                    path = msg['path']
                    cwd = msg['cwd']
                    argv = msg.get('argv', [])
                    env = msg.get('env', {})
                    encoding = msg.get('encoding', None)
                    print(f'[VSC] Executing script {path}')

                    pf = PythonFile(path, cwd, argv, env, encoding)

                    self.write_message({'type': MessageType.ServerReady})

                    debugpy.wait_for_client()
                    pf.exec()

                    self.write_message({'type': MessageType.DebugFinished})


                case _:
                    print(f'[VSC] Unknown message type {msg["type"]}')
                    self.write_message({'type': MessageType.Error, 'info': 'Unknown message type'})
        except Exception as e:
            info = f'{e.__class__.__name__}: {e}'
            self.write_message({'type': MessageType.Error, 'info': info})
            print(f'[!!!] {info}')

    def on_close(self) -> None:
        print('[VSC] Connect closed')


class Server(object):
    def __init__(self, config:Config):
        self.config = config
        self.app = tornado.web.Application([(r'/', _Server)])
        self.server = tornado.httpserver.HTTPServer(self.app)
        self.thread = threading.Thread(target=self._start)
        self.thread.daemon = True
        self.ioloop = None
        
    @property
    def running(self):
        return self.thread.is_alive()

    def start(self):
        '''start server'''
        if not self.running:
            if self.thread.ident is not None:
                # threads can only be started once
                self.thread = threading.Thread(target=self._start)
                self.thread.daemon = True
            self.thread.start()
        else:
            print('[VSC] server is already running')
    
    def _start(self):
        # if sys.version_info >= (3, 4):
        asyncio.set_event_loop(asyncio.new_event_loop())
        self.server.listen(address=self.config.host, port=self.config.port)
        self.ioloop = tornado.ioloop.IOLoop.current()
        self.ioloop.start()

    def stop(self):
        '''stop server'''
        global dbgsrv_running
        if not self.running:
            return
        # self.ioloop.stop()
        # self.ioloop = None
        self.ioloop.add_callback(self.ioloop.stop)
        # asyncio.new_event_loop().run_until_complete(self.server.close_all_connections())
        self.server.stop()
        self.thread.join()
        
        self.thread = threading.Thread(target=self._start)
        self.thread.daemon = True

        if dbgsrv_running:
            raise DebugServerCannotStopError('Debug server cannot be stopped currently.\ncheck here: https://github.com/microsoft/debugpy/issues/870')


idavscode/src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { WorkspaceFolder, DebugConfiguration, ProviderResult, CancellationToken } from 'vscode';
import * as WebSocket from 'ws';

const enum MessageType {
	startDebugServer = 'startDebugServer',
	stopDebugServer = 'stopDebugServer',
	stopServer = 'stopServer',
	executeScript = 'executeScript',

	serverReady = 'serverReady',
	debugServerReady = 'debugServerReady',
	debugFinished = 'debugFinished',
	error = 'error'
}

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
	// context.subscriptions.push(vscode.commands.registerCommand('idacode.connectToIDA', connectToIDA));
	const provider = new IDAPythonConfigurationProvider();
	context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider('idapython', provider));
}

// this method is called when your extension is deactivated
export function deactivate() { }



class IDAPythonConfigurationProvider implements vscode.DebugConfigurationProvider {

	provideDebugConfigurations?(folder: WorkspaceFolder | undefined, token?: CancellationToken): ProviderResult<DebugConfiguration[]> {
		return [
			{
				"name": "IDAPython debug",
				"type": "idapython",
				"request": "launch",
				"program": "${file}",
				"host": "localhost",
				"port": 5677,
				"timeout": 3000,
				"pythonPath": "${command:python.interpreterPath}",
				"logFile": "",
				"debugConfig": {
					"type": "python",
					"request": "attach",
					"justMyCode": true,
					"connect": {
						"port": 5678
					},
					"cwd": "${workspaceFolder}"
				}
			}
		];
	}

	/**
	 * Massage a debug configuration just before a debug session is being launched,
	 * e.g. add all missing attributes to the debug configuration.
	 */

	resolveDebugConfiguration(folder: WorkspaceFolder | undefined, config: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {
		if (!config.host || !config.port || !config.debugConfig || !config.debugConfig.connect || !config.debugConfig.connect.port) {
			return vscode.window.showInformationMessage("complete the debug config").then(_ => {
				return undefined;	// abort launch
			});
		}

		if (!config.program) { config.program = '${file}'; }
		if (!config.pythonPath) { config.pythonPath = '${command:python.interpreterPath}'; }
		if (config.timeout === undefined) { config.timeout = 3000; }
		config.debugConfig.type = 'python';
		config.debugConfig.request = 'attach';
		config.debugConfig.redirectOutput = true;
		config.debugConfig.showReturnValue = true;
		config.debugConfig.debugOptions = ["RedirectOutput", "ShowReturnValue"];
		if (!config.debugConfig.cwd) { config.debugConfig.cwd = '${workspaceFolder}'; }
		if (!config.debugConfig.connect.host) { config.debugConfig.connect.host = config.host; }

		return config;
	}

	resolveDebugConfigurationWithSubstitutedVariables(folder: WorkspaceFolder | undefined, config: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {

		return remoteIDAPythonExec(config);
	}
}


async function remoteIDAPythonExec(config: DebugConfiguration): Promise<DebugConfiguration | undefined> {
	try {
		await _remoteIDAPythonExec(config);
		vscode.commands.executeCommand('workbench.debug.action.focusRepl');
		return config.debugConfig;
	}
	catch (e: Error | string | any) {
		vscode.window.showErrorMessage(e);
		return undefined;	// abort launch
	}
}

function _remoteIDAPythonExec(config: DebugConfiguration): Promise<string> {
	let socket = new WebSocket(`ws://${config.host}:${config.port}/`);
	let timer: NodeJS.Timeout | undefined = undefined;

	const clearTimer = () => {
		if (timer) {
			clearTimeout(timer);
			timer = undefined;
		}
	};

	return new Promise((resolve, reject) => {

		timer = setTimeout(() => {
			socket.close();
			reject(new Error(`WebSocket connection timeout after ${config.timeout} ms`));
		}, config.timeout);

		socket.on('open', () => {
			clearTimer();
			socket.send(JSON.stringify({
				'type': MessageType.startDebugServer,
				'host': config.debugConfig.connect.host,
				'port': config.debugConfig.connect.port,
				'logfile': config.logFile,
				'pythonPath': config.pythonPath
			}));
		});

		socket.on('message', (data: WebSocket.Data) => {
			const json = JSON.parse(data.toString());
			switch (json.type) {
				case MessageType.debugServerReady:
					socket.send(JSON.stringify({
						'type': MessageType.executeScript,
						'path': config.program,
						'cwd': config.debugConfig.cwd,
						'argv': config.debugConfig.args,
						'env': config.debugConfig.env,
						// 'encoding': config.encoding
						// sadly, vscode seem to not provide api to get file encoding
					}));
					break;

				case MessageType.serverReady:
					resolve('ok');
					break;

				case MessageType.debugFinished:
					socket.close();
					vscode.debug.stopDebugging();
					break;

				case MessageType.error:
					reject(json.info);
					socket.close();
					break;
			}
		});

		socket.on('error', (err) => {
			clearTimer();
			reject(err);
		});
	});
}